### PR TITLE
Increase testing coverage

### PR DESCRIPTION
**Pull request by antalszava**
**Context**

Our codebase has lines that are not being covered by our tests. This may result in failing checks from Codecov, even when a PR author has not introduced any new uncovered changes.

**Changes**

Starts covering up the remaining non-covered lines of the codebase.

The following lines are not covered (based on running `make coverage` locally):
- [X] pennylane/_device.py                                                      308      7    98%   144, 148-149, 465, 1011, 1035, 1053
- [X] pennylane/_qubit_device.py                                                316     10    97%   620, 631, 641, 666, 810-814, 839-843
~- [ ] pennylane/beta/__init__.py                                                  1      1     0%   14~
- [x] pennylane/circuit_graph.py                                                168     12    93%   182-190, 407-409, 467
- [x] pennylane/configuration.py                                                 58      4    93%   61-63, 66
- [x] pennylane/devices/default_gaussian.py                                     292      1    99%   813 ~(PolyXP var; fix done, need a test)~
- [x] pennylane/devices/default_qubit_jax.py                                     89      3    97%   231-233
- [x] pennylane/devices/default_qubit_tf.py                                      62      5    92%   24, 30-31, 44-45
- [x] pennylane/devices/default_qubit_torch.py                                  101      8    92%   25-28, 255-258, 271
- [x] pennylane/gradients/parameter_shift_cv.py                                 245      2    99%   404, 479
- [x] pennylane/math/single_dispatch.py                                         259      7    97%   117, 307, 329-331, 389-390
- [x] pennylane/measurements.py                                                 237      1    99%   312
- [x] pennylane/numpy/tensor.py                                                  81      2    98%   181, 309
- [ ] pennylane/operation.py                                                    625    410    34%   177-190, 203-249, 265-272, 311-327, 330-333, 473-485, 500-503, 516-775, 779, 832-839, 847, 860, 866-868, 873, 876, 881, 883, 903-906, 914-915, 919, 924-925, 937, 944-947, 955-957, 976-1086, 1094-1101, 1103-1105, 1114, 1118, 1129-1173, 1176-1228, 1248-1250, 1260-1303, 1308, 1319-1340, 1353, 1356-1365, 1378, 1389, 1391, 1395-1397, 1423-1456, 1478, 1494-1522, 1525-1528, 1532, 1549, 1552-1600, 1608, 1617, 1621-1685, 1690, 1694-1695, 1700, 1703, 1705, 1711, 1713, 1721, 1724, 1731, 1741-1797, 1804-1807, 1813, 1816-1817, 1819, 1822, 1826, 1837-1849, 1852-1857, 1861, 1872, 1877-1929, 1934-1937, 1957-1968, 1970, 1973-2026, 2032-2077, 2105-2117, 2123-2130, 2134-2136, 2138-2139, 2146, 2155, 2157-2159, 2188-2201, 2222-2309, 2345-2402, 2405, 2407, 2413, 2422-2425, 2442-2445, 2449-2452, 2461-2462, 2464-2465
- [x] pennylane/optimize/rotoselect.py                                           46      2    96%   142-143
- [x] pennylane/qaoa/cycle.py                                                   180      2    99%   492, 540
- [ ] pennylane/qnn/keras.py                                                     74      4    95%   27-33
- [ ] pennylane/qnn/torch.py                                                     61      4    93%   27-33
- [ ] pennylane/queuing.py                                                       85      1    99%   209
- [ ] pennylane/tape/tape.py                                                    427      5    99%   350-352, 686-687
--------
Christina (https://github.com/PennyLaneAI/pennylane/pull/2537):
- [x] pennylane/drawer/mpldrawer.py 213 3 99% 23-24, 250
- [x] pennylane/fourier/coefficients.py 31 1 97% 164
- [x] pennylane/fourier/visualize.py 154 2 99% 24-25
- [x] pennylane/ops/functions/generator.py 70 1 99% 123
- [x] pennylane/ops/qubit/hamiltonian.py 212 2 99% 624-625
- [x] pennylane/ops/qubit/matrix_ops.py 105 1 99% 409

--------
Romain (https://github.com/PennyLaneAI/pennylane/pull/2539):
- [x] pennylane/templates/embeddings/amplitude.py                                50      1    98%   221
- [x] pennylane/transforms/commutation_dag.py                                   323      1    99%   694
- [x] pennylane/transforms/metric_tensor.py                                     182      1    99%   333
- [x] pennylane/transforms/optimization/merge_rotations.py                       42      2    95%   144, 153
- [x] pennylane/transforms/optimization/pattern_matching.py                     620      2    99%   1232-1233
- [x] pennylane/utils.py                                                        148      3    98%   197-198, 414

### CODE REVIEW

File: pennylane/circuit_graph.py [hunk 1]
```diff
@@ -394,20 +394,6 @@ def nodes_between(self, a, b):
         B.add(b)
         return A & B
 
-    def invisible_operations(self):
```
Comment (antalszava) [inline pennylane/circuit_graph.py:397]: Added in https://github.com/PennyLaneAI/pennylane/pull/488, but not used anymore.

**Comment by antalszava** [comment]
[sc-17362]

**Comment by rmoyard** [comment]
@antalszava  I will take care of  

1.   pennylane/templates/embeddings/amplitude.py 50 1 98% 221
2.  pennylane/transforms/commutation_dag.py 323 1 99% 694
3.  pennylane/transforms/metric_tensor.py 182 1 99% 333
4.  pennylane/transforms/optimization/merge_rotations.py 42 2 95% 144, 153
5.  pennylane/transforms/optimization/pattern_matching.py 620 2 99% 1232-1233
6.  pennylane/utils.py 148 3 98% 197-198, 414

**Comment by antalszava** [comment]
Note: going forward with this PR, the originally found list is not entirely compete. Potentially, there may have been other deviations too since the generation of that list.

File: pennylane/circuit_graph.py (hunk 1 above)
Comment (timmysilv) [reply pennylane/circuit_graph.py:394]: does it not require any deprecation?

**Review by timmysilv — COMMENTED** [review]
all looks good! I'm a little scared to ask but because it's been so long, do you know if you're adding any redundant tests? I'd say this should be merged in general because it's not worth waiting until every single line is covered. we can always open another PR. curious about the deprecation comment but otherwise, more tests generally sounds good!

**Comment by antalszava** [comment]
> all looks good! I'm a little scared to ask but because it's been so long, do you know if you're adding any redundant tests?

@timmysilv thanks a lot for this! I actually found a case that has been added in the meantime, the rest should be fine.

**Review by timmysilv — APPROVED** [review]
:shipit: :shipit: :shipit:
